volume/thin_pool_watcher.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package volume
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// ThinPoolWatcher maintains a cache of device name -> usage stats for a devicemapper thin-pool using thin_ls.
+type ThinPoolWatcher struct {
+	poolName string
+	lock     *sync.RWMutex
+	cache    map[string]uint64
+}
+
+// NewThinPoolWatcher returns a new ThinPoolWatcher for the given devicemapper thin pool name.
+func NewThinPoolWatcher(poolName string) *ThinPoolWatcher {
+	return &ThinPoolWatcher{poolName, &sync.RWMutex{}, make(map[string]uint64)}
+}
+
+func (w *ThinPoolWatcher) Start() {
+	// NOTE: placeholder code, I intend to find a way to reuse kube wait.Forever here
+	for {
+		select {
+		case <-time.After(0):
+			w.Refresh()
+		}
+	}
+}
+
+// GetUsage gets the cached usage value of the given device.
+func (w *ThinPoolWatcher) GetUsage(deviceId string) (uint64, error) {
+	w.lock.RLock()
+	defer w.lock.RUnlock()
+	v, ok := w.cache[deviceId]
+	if !ok {
+		return 0, fmt.Errorf("No cached value for usage of device %v", deviceId)
+	}
+
+	return v, nil
+}
+
+// Refresh performs a `thin_ls` of the pool being watched and refreshes the
+// cached data with the result.
+func (w *ThinPoolWatcher) Refresh() {
+	output, err := doThinLs(w.poolName)
+	if err != nil {
+		glog.Errorf("Unable to get thin-pool usage for pool %v: %v", w.poolName, err)
+		return
+	}
+
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	w.cache = parseThinLsOutput(output)
+}
+
+// doThinLs handles obtaining the output of thin_ls for the given pool name; it:
+//
+// 1. Reserves a metadata snapshot for the pool
+// 2. Runs thin_ls against that snapshot
+// 3. Releases the snapshot
+func doThinLs(poolName string) ([]byte, error) {
+	// (1)
+	// NOTE: "0" in the call below is for the 'sector' argument to 'dmsetup message'.  It's not needed for thin pools.
+	if _, err := exec.Command("dmsetup", "message", poolName, "0", "reserve_metadata_snap").Output(); err != nil {
```

Comment:
@ncdc, yeah, this isn't intended to be the final factoring -- just trying
to get the rough factoring w/in cadvisor correct before getting the little
details right

On Thu, Apr 21, 2016 at 1:53 PM, Andy Goldstein <notifications@github.com>
wrote:

> In volume/thin_pool_watcher.go
> <https://github.com/google/cadvisor/pull/1204#discussion_r60626204>:
>
> > +
> > +	w.lock.Lock()
> > +	defer w.lock.Unlock()
> > +
> > +	w.cache = parseThinLsOutput(output)
> > +}
> > +
> > +// doThinLs handles obtaining the output of thin_ls for the given pool name; it:
> > +//
> > +// 1. Reserves a metadata snapshot for the pool
> > +// 2. Runs thin_ls against that snapshot
> > +// 3. Releases the snapshot
> > +func doThinLs(poolName string) ([]byte, error) {
> > +	// (1)
> > +	// NOTE: "0" in the call below is for the 'sector' argument to 'dmsetup message'.  It's not needed for thin pools.
> > +	if _, err := exec.Command("dmsetup", "message", poolName, "0", "reserve_metadata_snap").Output(); err != nil {
>
> I'd recommend creating an interface with functions such as
> reserveMetadataSnap and releaseMetadataSnap so this can be unit tested.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/cadvisor/pull/1204/files/f984134c77bde6df61048b98d9dce966aa60b513#r60626204>
>